### ISSUE TITLE
Table with unique index

### ISSUE THREAD

**Issue by vmihailenco** [reorg/pg_repack#70]
pg_repack sometimes can't repack relatively busy table with unique index

```
ERROR: query failed: ERROR:  duplicate key value violates unique constraint "index_287502"
DETAIL:  Key (group_id, hash)=(1591954043305993739, 5338517406217073467) already exists.
CONTEXT:  SQL statement "INSERT INTO repack.table_17593 VALUES ($1.*)"
DETAIL: query was: SELECT repack.repack_apply($1, $2, $3, $4, $5, $6)
```

What is the right strategy in such situation? Just re-run pg_repack?

**Comment by kotsachin** [comment]
Can you tell how you produced above error?
I would like to try.

**Comment by vmihailenco** [comment]
1. Create main.go from following script https://gist.github.com/vmihailenco/2ab6aaa5b8143c4b1b67

2. Start it

```
go run main.go
2016/02/09 11:34:52 ready...
```

3. Run pg_repack

```
sudo -u postgres /usr/lib/postgresql/9.5/bin/pg_repack postgres -t repack_issue_70
INFO: repacking table "repack_issue_70"
ERROR: query failed: ERROR:  duplicate key value violates unique constraint "index_67706"
DETAIL:  Key (hash)=(23) already exists.
CONTEXT:  SQL statement "INSERT INTO repack.table_67700 VALUES ($1.*)"
DETAIL: query was: SELECT repack.repack_apply($1, $2, $3, $4, $5, $6)
```

PS You will need PostgreSQL 9.5 for upsert functionality.

**Comment by amitlan** [comment]
Hmm, I wonder if someone has verified pg_repack functionality with concurrent upserts in 9.5. I'm not entirely sure but there might some race conditions involved...

**Comment by vmihailenco** [comment]
This is version without upsert that can trigger this bug: https://gist.github.com/vmihailenco/78768e6eaa93aabd5888

**Comment by j16sdiz** [comment]
@vmihailenco I can confirm this as well.

**Comment by vmihailenco** [comment]
Gentle ping. Any chance pg_repack can be fixed to continue working in such cases?

**Comment by schmiddy** [comment]
I was able to reproduce the bug with @vmihailenco's [testcase](https://gist.github.com/vmihailenco/78768e6eaa93aabd5888). Thanks for putting it together, it's a very interesting testcase to play with.

I'm sorry I haven't had much time recently to dig more into this, but I have at least a rough guess of what's going on. Basically, `pg_repack` takes care to ensure that every INSERT, UPDATE, and DELETE applied on the target table also gets applied on the repacked version during the rebuild period. But particularly for this testcase, the ordering of changes applied by concurrent transactions is crucially important . If `pg_repack` doesn't apply its changes in exactly the order they were originally applied, we can run into trouble, and I believe that's exactly what's happening here.

The ordering that we use really depends on two assumptions: first, that the bigserial `id` of the repack.log_ table can faithfully serialize and preserve the order of the original INSERTS / UPDATES / DELETEs, and second, that rows in the log table become visible post-commit to the transaction that `repack_apply()` is using, also in the same order. I have a feeling that this testcase is exposing a hole in one or both of those assumptions :-(

As for what to do about this, I'm not really sure. One thought that occurred to me was that we should really be checking that every UPDATE and DELETE we apply on the repacked table affects exactly one row, and fail loudly if it doesn't (or try to defer the UPDATE or DELETE if it affected zero rows?). Other ideas are welcome.

**Comment by vmihailenco** [comment]
My first idea is that we can probably create unique index (or all indexes?) after repack log was replayed. What do you think? This at least should lower the chance of conflict.

Also looking at log

```
LOG: (query) SELECT repack.repack_apply($1, $2, $3, $4, $5, $6)
LOG: 	(param:0) = SELECT * FROM repack.log_65424 ORDER BY id LIMIT $1
LOG: 	(param:1) = INSERT INTO repack.table_65424 VALUES ($1.*)
LOG: 	(param:2) = DELETE FROM repack.table_65424 WHERE (id) = ($1.id)
LOG: 	(param:3) = UPDATE repack.table_65424 SET (id, total, additions, deletions) = ($2.id, $2.total, $2.additions, $2.deletions) WHERE (id) = ($1.id)
LOG: 	(param:4) = DELETE FROM repack.log_65424 WHERE id IN (
LOG: 	(param:5) = 0
```

it seems that first repack replays insert, then delete and then update, which explain why there is a conflict.

>One thought that occurred to me was that we should really be checking that every UPDATE and DELETE we apply on the repacked table affects exactly one row, and fail loudly if it doesn't (or try to defer the UPDATE or DELETE if it affected zero rows?)

AFAIK my testcase affects exactly one row so your suggestion will not fix it...

**Comment by MasahikoSawada** [comment]
> I'm sorry I haven't had much time recently to dig more into this, but I have at least a rough guess of what's going on. Basically, pg_repack takes care to ensure that every INSERT, UPDATE, and DELETE applied on the target table also gets applied on the repacked version during the rebuild period. But particularly for this testcase, the ordering of changes applied by concurrent transactions is crucially important . If pg_repack doesn't apply its changes in exactly the order they were originally applied, we can run into trouble, and I believe that's exactly what's happening here.

This is the same as the result of what I investigated about this issue. A insertion to repack.log_XXX by a transaction can be happened before an another preceding transaction inserts the operation log to repack.log_XXX. Suppose the table with primary key has data 1. And Tx1 and Tx2 are executed concurrently on that table. 

+ Tx1: DELETE 1;
+ Tx2: INSERT 1;

If these transactions are executed in correct order Tx1 -> Tx2 on the original table and are inserted to repack.log_XXX in the same order, repack_apply should be done successfully. But it's possible that transactions are executed in order Tx1->Tx2 on original table but are inserted in order Tx2->Tx1. In this case, repack_apply will failed due to unique constraint violation.

> My first idea is that we can probably create unique index (or all indexes?) after repack log was replayed. What do you think? 

I think that this could solve this issue but it makes swapping take long time, which means duration of holding the exclusive lock can be long.  Especially if very large table with indexes, it would be terrible.

Another idea I came up with is to defer the checking unique constraint during applying operation log. We create the repacked table(repack.table_XXX) with DEFERRABLE option when CREATE TABLE and then execute SET CONSTRAINTS DEFERED at starting of apply phase. Checking unique constraint will be invoked after all operation logs are applied and pg_repack emits error if unique constraint violation. The big point I'm concerned is that we can ensure that data of repacked table is always consistent with the original table if unique constraint violation is not occurred during the apply phase. Thought?

**Comment by bradley-fleming** [comment]
We just ran into this same bug repacking a table that does use upsert.

**Comment by MasahikoSawada** [comment]
Yes, v1.3.4 cannot handle the situation where upsert is executed during repacking table. Commit 5781aabb61e50d5ab3aae0561ae4dcb3123e5525 fixed this issue and v1.4.0, which incorporates this fix, will be release soon.

**Comment by MichaelDBA** [comment]
Closing since fixed in previous release, v1.4.0.
